### PR TITLE
Adding new test that tests fastboot with zero lamport accounts

### PR DESCRIPTION
**Pull request by roryharr**
#### Problem
- Currently no testing that tests fastboot with zero lamport accounts present. This specific scenario will fail with obsolete accounts enabled if they are not preserved. It's also useful to ensure it doesn't fail without obsolete accounts

#### Summary of Changes
- Add the test. 
- Set up as a test case so obsolete accounts can be added when they support fastboot
- Failure logs below (with obsolete accounts enabled)

```
mac:runtime roryharris$ cargo test test_restore_snapshots_dir_handle_zero_lamport_accounts
   Compiling solana-runtime v3.1.0 (/Users/roryharris/src/agave/runtime)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 3.78s
     Running unittests src/lib.rs (/Users/roryharris/src/agave/target/debug/deps/solana_runtime-48f16474e929af66)

running 2 tests
test snapshot_bank_utils::tests::test_restore_snapshots_dir_handle_zero_lamport_accounts::markobsoleteaccounts_enabled_expects ... FAILED
test snapshot_bank_utils::tests::test_restore_snapshots_dir_handle_zero_lamport_accounts::markobsoleteaccounts_disabled_expects ... ok

failures:

---- snapshot_bank_utils::tests::test_restore_snapshots_dir_handle_zero_lamport_accounts::markobsoleteaccounts_enabled_expects stdout ----

thread 'snapshot_bank_utils::tests::test_restore_snapshots_dir_handle_zero_lamport_accounts::markobsoleteaccounts_enabled_expects' panicked at runtime/src/snapshot_bank_utils.rs:2063:10:
called `Result::unwrap()` on an `Err` value: AccountsFileError(AppendVecError(IncorrectLayout("/var/folders/ds/23ynjk1d04v97trb0zxq99600000gn/T/.tmpntxpCN/run/0.0")))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    snapshot_bank_utils::tests::test_restore_snapshots_dir_handle_zero_lamport_accounts::markobsoleteaccounts_enabled_expects

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 532 filtered out; finished in 0.64s
```

Fixes #

### CODE REVIEW

File: runtime/src/snapshot_bank_utils.rs [hunk 1]
```diff
@@ -1966,6 +1966,109 @@ mod tests {
...
+    ///
+    /// slot 1:
+    ///     - send all lamports to account2 from account1 to make account1 zero lamport
+    /// slot 2:
+    ///     - send all lamports from account2 to the mint to make account2 zero lamport
+    ///
+    /// If zero lamport accounts are not handled correctly, Account1 or Account2 will come back
+    /// failing the test
+    #[test_case(MarkObsoleteAccounts::Disabled)]
+    fn test_restore_snapshots_dir_handle_zero_lamport_accounts(
```
Comment (roryharr) [inline runtime/src/snapshot_bank_utils.rs]: This is intended to be an targeted test. For more general testing of loading from fast-boot, test_bank_from_snapshot_dir could be modified to support some test case configuration (number of slots, number of accounts etc...)

File: runtime/src/snapshot_bank_utils.rs [hunk 2]
```diff
@@ -1827,7 +1827,7 @@ mod tests {
         }
     }
 
-    /// Test that snapshots correctly handle zero lamport accounts
+    /// Test that snapshots correctly handles zero lamport accounts
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: I think the original version is grammatically correct; "snapshots handle".


```suggestion
    /// Test that snapshots correctly handle zero lamport accounts
```

**Review by brooksprumo — COMMENTED** [review]
This test doesn't fastboot though, so it's a little confusing what the intent is.

I think the intent is that calling `bank_from_snapshot_dir()` here is meant to mean "fastboot", but that's only half the picture; just the load-side

Fastboot requires special setup on the store-side as well. Meaning, when generating a snapshot that we flush and hard link storages, plus write the various "snapshot complete" files. And now serializing the obsolete accounts.

`bank_from_snapshot_dir()` is the load-side. Fastboot must call this fn, yes. It was also intended to be the path for loading from snapshot archives as well.

Maybe it'll make more sense by adding in the store-side of fastboot as well?

File: runtime/src/snapshot_bank_utils.rs (hunk 2 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Haha. I meant to change this comment where it is not grammatically correct. Will fix. 
 /// Test that restoring from snapshot dir rather than archives correctly handle zero lamport accounts

Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Can it be in a different PR though?

**Comment by roryharr** [comment]
> This test doesn't fastboot though, so it's a little confusing what the intent is.
> 
> I think the intent is that calling `bank_from_snapshot_dir()` here is meant to mean "fastboot", but that's only half the picture; just the load-side
> 
> Fastboot requires special setup on the store-side as well. Meaning, when generating a snapshot that we flush and hard link storages, plus write the various "snapshot complete" files. And now serializing the obsolete accounts.
> 
> `bank_from_snapshot_dir()` is the load-side. Fastboot must call this fn, yes. It was also intended to be the path for loading from snapshot archives as well.
> 
> Maybe it'll make more sense by adding in the store-side of fastboot as well?

All of the store side code is called if true is passed into bank_to_full_archive_with. The true is the parameter should_flush_and_hard_link_storages which ends up calling the same code as teardown()

This is how the existing test_bank_from_snapshot_dir works, which is what I used as a reference.

File: runtime/src/snapshot_bank_utils.rs (hunk 2 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: No I meant: I looked at my PR and saw a line I added was grammatically incorrect. 
I did a quick search and replace, and replaced the wrong one. Yes, I will revert this change. I had no intention of making it.

**Review by HaoranYi — DISMISSED** [review]
LGTM.

File: runtime/src/snapshot_bank_utils.rs [hunk 3]
```diff
@@ -1966,6 +1966,110 @@ mod tests {
         assert_eq!(*bank3, deserialized_bank);
     }
 
+    /// Test that restoring from snapshot dir rather than archives correctly handles zero lamport
```
Comment (HaoranYi) [inline runtime/src/snapshot_bank_utils.rs]: 👍 Good documentation explaining the flow

**Comment by roryharr** [comment]
> > This test doesn't fastboot though, so it's a little confusing what the intent is.
> > I think the intent is that calling `bank_from_snapshot_dir()` here is meant to mean "fastboot", but that's only half the picture; just the load-side
> > Fastboot requires special setup on the store-side as well. Meaning, when generating a snapshot that we flush and hard link storages, plus write the various "snapshot complete" files. And now serializing the obsolete accounts.
> > `bank_from_snapshot_dir()` is the load-side. Fastboot must call this fn, yes. It was also intended to be the path for loading from snapshot archives as well.
> > Maybe it'll make more sense by adding in the store-side of fastboot as well?
> 
> All of the store side code is called if true is passed into bank_to_full_archive_with. The true is the parameter should_flush_and_hard_link_storages which ends up calling the same code as teardown()
> 
> This is how the existing test_bank_from_snapshot_dir works, which is what I used as a reference.

Discussed and added a comment to make this more clear. 
Probably makes sense to add an enum in the future and rename it. Adding to the list of debt.

**Comment by brooksprumo** [comment]
> All of the store side code is called if true is passed into bank_to_full_archive_with. The true is the parameter should_flush_and_hard_link_storages which ends up calling the same code as teardown()

Ah, thank you, I missed that!

File: runtime/src/snapshot_bank_utils.rs [hunk 4]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
...
+    ///
+    /// slot 1:
+    ///     - send all lamports to account2 from account1 to make account1 zero lamport
+    /// slot 2:
+    ///     - send all lamports from account2 to the mint to make account2 zero lamport
+    ///
+    /// If zero lamport accounts are not handled correctly, Account1 or Account2 will come back
+    /// failing the test
+    #[test_case(MarkObsoleteAccounts::Disabled)]
+    fn test_restore_snapshots_dir_handle_zero_lamport_accounts(
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Can we rename this test to say "fastboot" instead? Maybe:

```suggestion
    fn test_fastboot_handles_zero_lamport_accounts(
```

(and also the comments above).

File: runtime/src/snapshot_bank_utils.rs [hunk 5]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
         assert_eq!(*bank3, deserialized_bank);
     }
 
+    /// Test that restoring from snapshot dir rather than archives correctly handles zero lamport
+    /// accounts
+    ///
+    /// slot 0:
+    ///     - send some lamports to Account2 (from mint) to bring it to life
+    ///     - send some lamports to Account1 (from mint) to bring it to life
+    ///
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs:1974]: nit: remove newline here to be consistent with slot 1 and 2?

```suggestion
```

File: runtime/src/snapshot_bank_utils.rs [hunk 6]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
...
+        let bank_test_config = BankTestConfig {
+            accounts_db_config: AccountsDbConfig {
+                mark_obsolete_accounts,
+                ..ACCOUNTS_DB_CONFIG_FOR_TESTING
+            },
+        };
+
+        let bank0 = Bank::new_with_config_for_tests(&genesis_config, bank_test_config);
+        let (bank0, bank_forks) = Bank::wrap_with_bank_forks_for_tests(bank0);
+        // Transport some money from mint to key1 and key2
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: nit: IMO this wording a little strange ("transport" vs transfer, and "money" vs lamports). But also doesn't really add any info beyond what the next two lines of code do. Maybe just remove?


```suggestion
```

File: runtime/src/snapshot_bank_utils.rs [hunk 7]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
...
+        bank0.transfer(lamports, &mint, &key2.pubkey()).unwrap();
+        bank0.transfer(lamports, &mint, &key1.pubkey()).unwrap();
+        bank0.fill_bank_with_ticks_for_tests();
+
+        // Squash and root bank0 to ensure slot 0. This ensures slot0 data is not cleaned
+        // before being written to storage.
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Seems like this is mid-update?

File: runtime/src/snapshot_bank_utils.rs [hunk 8]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
...
+            false,
+            ACCOUNTS_DB_CONFIG_FOR_TESTING,
+            None,
+            Arc::default(),
+        )
+        .unwrap();
+
+        // Ensure both accounts are still zero lamport
+        assert_eq!(deserialized_bank.get_balance(&key1.pubkey()), 0,);
+        assert_eq!(deserialized_bank.get_balance(&key2.pubkey()), 0,);
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: nit nit: I wish formatting of macros didn't leave the trailing comma. Mind removing?

```suggestion
        assert_eq!(deserialized_bank.get_balance(&key1.pubkey()), 0);
        assert_eq!(deserialized_bank.get_balance(&key2.pubkey()), 0);
```

File: runtime/src/snapshot_bank_utils.rs [hunk 9]
```diff
@@ -1966,6 +1966,112 @@ mod tests {
...
+    ///     - send some lamports to Account1 (from mint) to bring it to life
+    ///
+    /// slot 1:
+    ///     - send all lamports to account2 from account1 to make account1 zero lamport
+    /// slot 2:
+    ///     - send all lamports from account2 to the mint to make account2 zero lamport
+    ///
+    /// If zero lamport accounts are not handled correctly, Account1 or Account2 will come back
+    /// failing the test
+    #[test_case(MarkObsoleteAccounts::Disabled)]
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs:1983]: I assume once serializing of the obsolete accounts is implemented, then this test with add another test case for `MarkObsoleteAccounts::Enabled`?

**Review by brooksprumo — COMMENTED** [review]
Some nits. Looks good overall!

File: runtime/src/snapshot_bank_utils.rs (hunk 9 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs:1983]: Yup, thats the plan!

File: runtime/src/snapshot_bank_utils.rs (hunk 4 above)
Comment (roryharr) [reply runtime/src/snapshot_bank_utils.rs]: Sure

File: runtime/src/snapshot_bank_utils.rs (hunk 7 above)
Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Maybe this was the intention:

```suggestion
        // Squash and flush bank0 to ensure slot 0 data is not cleaned
        // before being written to storage.
```

File: runtime/src/snapshot_bank_utils.rs [hunk 10]
```diff
@@ -1966,6 +1966,110 @@ mod tests {
...
+    ///
+    /// slot 2:
+    ///     - send all lamports from account2 to the mint to make account2 zero lamport
+    ///
+    /// If zero lamport accounts are not handled correctly, Account1 or Account2 will come back
+    /// failing the test
+    #[test_case(MarkObsoleteAccounts::Disabled)]
+    fn test_fastboot_handle_zero_lamport_accounts(
+        mark_obsolete_accounts: MarkObsoleteAccounts,
+    ) {
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Looks like cargo-fmt wants this to be formatted differently.

**Review by brooksprumo — COMMENTED** [review]
Looks like

File: runtime/src/snapshot_bank_utils.rs [hunk 11]
```diff
@@ -1966,6 +1966,108 @@ mod tests {
...
+            },
+        };
+
+        let bank0 = Bank::new_with_config_for_tests(&genesis_config, bank_test_config);
+        let (bank0, bank_forks) = Bank::wrap_with_bank_forks_for_tests(bank0);
+        bank0.transfer(lamports, &mint, &key2.pubkey()).unwrap();
+        bank0.transfer(lamports, &mint, &key1.pubkey()).unwrap();
+        bank0.fill_bank_with_ticks_for_tests();
+
+        // Squash and root bank0. This ensures slot0 data is not cleaned before being written to
```
Comment (brooksprumo) [inline runtime/src/snapshot_bank_utils.rs]: Squashing roots the bank, which is why I removed "and root" in https://github.com/anza-xyz/agave/pull/8297#discussion_r2403148644.

Could also be "Squash to root bank0".

Or could be updated in a follow up PR.

File: runtime/src/snapshot_bank_utils.rs (hunk 7 above)
Comment (brooksprumo) [reply runtime/src/snapshot_bank_utils.rs]: Continued here: https://github.com/anza-xyz/agave/pull/8297#discussion_r2407177363.

**Review by brooksprumo — COMMENTED** [review]
Let's resolve https://github.com/anza-xyz/agave/pull/8297/files#r2407177363 and then it'll be good to go.

**Review by brooksprumo — APPROVED** [review]
:shipit:
